fe/fe-core/src/main/java/com/starrocks/catalog/ForeignKeyConstraint.java

```diff
@@ -0,0 +1,129 @@
+            "((\\.?\\w+)+)\\s*\\(((,?\\s*\\w+\\s*)+)\\)";
+    public static final Pattern FOREIGN_KEY_PATTERN = Pattern.compile(FOREIGN_KEY_REGEX);
+    // table with primary key or unique key
+    // if parent table is dropped, the foreign key is not dropped cascade now.
+    private final BaseTableInfo parentTableInfo;
+
+    // here id is preferred, but meta of column does not have id.
+    // have to use name here, so column rename is not supported
+    // eg: [column1 -> column1', column2 -> column2']
+    private final List<Pair<String, String>> columnRefPairs;
+
+    public ForeignKeyConstraint(
+            BaseTableInfo parentTableInfo,
+            List<Pair<String, String>> columnRefPairs) {
+        this.parentTableInfo = parentTableInfo;
+        this.columnRefPairs = columnRefPairs;
+    }
+
+    public BaseTableInfo getParentTableInfo() {
+        return parentTableInfo;
+    }
+
+    public List<Pair<String, String>> getColumnRefPairs() {
+        return columnRefPairs;
+    }
+
+    // for default catalog, the format is: (column1, column2) REFERENCES default_catalog.dbid.tableid(column1', column2')
+    // for other catalog, the format is: (column1, column2) REFERENCES default_catalog.dbname.tablename(column1', column2')
+    @Override
+    public String toString() {
+        if (parentTableInfo == null || columnRefPairs == null) {
+            return "";
+        }
+        StringBuilder sb = new StringBuilder();
+        sb.append("(");
+        String baseColumns = Joiner.on(",").join(columnRefPairs.stream().map(pair -> pair.first).collect(Collectors.toList()));
+        sb.append(baseColumns);
+        sb.append(")");
+        sb.append(" REFERENCES ");
+        sb.append(parentTableInfo.toString());
+        sb.append("(");
+        String parentColumns = Joiner.on(",").join(columnRefPairs.stream().map(pair -> pair.second).collect(Collectors.toList()));
+        sb.append(parentColumns);
+        sb.append(")");
+        return sb.toString();
+    }
+
+    // for default catalog, the format is: (column1, column2) REFERENCES default_catalog.dbid.tableid(column1', column2')
+    // for other catalog, the format is: (column1, column2) REFERENCES default_catalog.dbname.tablename(column1', column2')
+    public static List<ForeignKeyConstraint> parse(String foreignKeyConstraintDescs) {
+        if (Strings.isNullOrEmpty(foreignKeyConstraintDescs)) {
+            return null;
+        }
+        String[] constraintArray = foreignKeyConstraintDescs.split(";");
+        List<ForeignKeyConstraint> foreignKeyConstraints = Lists.newArrayList();
+        for (String constraintDesc : constraintArray) {
+            if (Strings.isNullOrEmpty(constraintDesc)) {
+                continue;
+            }
+            Matcher foreignKeyMatcher = FOREIGN_KEY_PATTERN.matcher(constraintDesc);
+            if (!foreignKeyMatcher.find()) {
+                LOG.warn("invalid constraint:{}", foreignKeyConstraintDescs);
```

Comment:
Why not throw exception?

Reply:
the pattern has be checked in PropertyAnalyzer and here just a protection coding. and it is used when log replay, here just ignore it if something goes wrong here to make replay more robust.

Reply:
it has no chance to go here now.